google/cloud/bigtable/data/_metrics/data_model.py

```diff
@@ -0,0 +1,495 @@
+    """Enum for the state of the active operation."""
+
+    CREATED = 0
+    ACTIVE_ATTEMPT = 1
+    BETWEEN_ATTEMPTS = 2
+    COMPLETED = 3
+
+
+@dataclass(frozen=True)
+class CompletedAttemptMetric:
+    """
+    An immutable dataclass representing the data associated with a
+    completed rpc attempt.
+
+    Operation-level fields (eg. type, cluster, zone) are stored on the
+    corresponding CompletedOperationMetric or ActiveOperationMetric object.
+    """
+
+    duration_ns: int
+    end_status: StatusCode
+    first_response_latency_ns: int | None = None
+    gfe_latency_ns: int | None = None
+    application_blocking_time_ns: int = 0
+    backoff_before_attempt_ns: int = 0
+    grpc_throttling_time_ns: int = 0
+
+
+@dataclass(frozen=True)
+class CompletedOperationMetric:
+    """
+    An immutable dataclass representing the data associated with a
+    completed rpc operation.
+
+    Attempt-level fields (eg. duration, latencies, etc) are stored on the
+    corresponding CompletedAttemptMetric object.
+    """
+
+    op_type: OperationType
+    duration_ns: int
+    completed_attempts: list[CompletedAttemptMetric]
+    final_status: StatusCode
+    cluster_id: str
+    zone: str
+    is_streaming: bool
+    flow_throttling_time_ns: int = 0
+
+
+@dataclass
+class ActiveAttemptMetric:
+    """
+    A dataclass representing the data associated with an rpc attempt that is
+    currently in progress. Fields are mutable and may be optional.
+    """
+
+    # keep monotonic timestamps for active attempts
+    start_time_ns: int = field(default_factory=time.monotonic_ns)
+    # the time it takes to recieve the first response from the server, in nanoseconds
+    # currently only tracked for ReadRows
+    first_response_latency_ns: int | None = None
+    # the time taken by the backend, in nanoseconds. Taken from response header
+    gfe_latency_ns: int | None = None
+    # time waiting on user to process the response, in nanoseconds
+    # currently only relevant for ReadRows
+    application_blocking_time_ns: int = 0
+    # backoff time is added to application_blocking_time_ns
+    backoff_before_attempt_ns: int = 0
+    # time waiting on grpc channel, in nanoseconds
+    # TODO: capture grpc_throttling_time
+    grpc_throttling_time_ns: int = 0
+
+
+@dataclass
+class ActiveOperationMetric:
+    """
+    A dataclass representing the data associated with an rpc operation that is
+    currently in progress. Fields are mutable and may be optional.
+    """
+
+    op_type: OperationType
+    backoff_generator: BackoffGenerator | None = None
+    # keep monotonic timestamps for active operations
+    start_time_ns: int = field(default_factory=time.monotonic_ns)
+    active_attempt: ActiveAttemptMetric | None = None
+    cluster_id: str | None = None
+    zone: str | None = None
+    completed_attempts: list[CompletedAttemptMetric] = field(default_factory=list)
+    is_streaming: bool = False  # only True for read_rows operations
```

Comment:
Thats not entirely true, it would also be true for CDC if we were to support that in this client and we have a couple of other features that would set this to true. I would remove the comment so it doesnt get stale